include/openssl/ssl.h.in

```diff
@@ -1754,7 +1754,7 @@ __owur X509 *SSL_get0_peer_certificate(const SSL *s);
 __owur X509 *SSL_get1_peer_certificate(const SSL *s);
 /* Deprecated in 3.0.0 */
 #  ifndef OPENSSL_NO_DEPRECATED_3_0
-#   define SSL_get_peer_certificate SSL_get1_peer_certificate
+#   define SSL_get_peer_certificate(s) (OSSL_DEPRECATED_PRAGMA("SSL_get_peer_certificate is deprecated, use SSL_get1_peer_certificate instead") SSL_get1_peer_certificate(s))
```

Comment:
This is fairly ugly. Could the OSSL_DEPRECATED_PRAGMA be defined in a way that would allow following?

```suggestion
#   define SSL_get_peer_certificate(s) OSSL_DEPRECATED_PRAGMA(SSL_get_peer_certificate, SSL_get1_peer_certificate)(s)
```

Reply:
As of [0] "The message contained in the pragma must be a single string literal." and with concatenating strings this will be no "single string literal" anymore.

Now the question becomes: what is uglier? Having this or no warnings on usage of deprecated macros?

[0] https://gcc.gnu.org/onlinedocs/cpp/Pragmas.html

Reply:
I am not convinced this is correct.  A constructed string literal is still a string literal. Indeed the documentation shows use of `#x`, so why not `#x #y` or `##x "foo"` ...
```
Its syntax is _Pragma (string-literal), where string-literal can be either a normal or
wide-character string literal. It is destringized, by replacing all ‘\\’ with a single
‘\’ and all ‘\"’ with a ‘"’. The result is then processed as if it had appeared as the
right hand side of a ‘#pragma’ directive. For example,

    _Pragma ("GCC dependency \"parse.y\"")

has the same effect as #pragma GCC dependency "parse.y". The same effect
could be achieved using macros, for example

    #define DO_PRAGMA(x) _Pragma (#x)
    DO_PRAGMA (GCC dependency "parse.y")
```

Reply:
Please fix it then for me, I couldn't do it.

Reply:
> Please fix it then for me, I couldn't do it.

... at least not in a way that the generated warning doesn't look as ugly as
```
warning: "SSL_get_peer_certificate" "is deprecated, use " "SSL_get1_peer_certificate" " instead"
```

Reply:
@vdukhovni anything to add?

Reply:
Below is a definition that meets the suggest from @t8m. Run this through `gcc -E` and see that it works correctly.
```c
#define OSSL_PRAGMA_WRAP(msg) _Pragma(#msg)
#define OSSL_PRAGMA_GCC_WARN(msg) OSSL_PRAGMA_WRAP(GCC warning # msg)
#define OSSL_DEPRECATED_INSTEAD_USE(f1, f2, ...) OSSL_PRAGMA_GCC_WARN(f1 is deprecated use f2 instead) f2(__VA_ARGS__)
#define SSL_get_peer_certificate(s) OSSL_DEPRECATED_INSTEAD_USE(SSL_get_peer_certificate, SSL_get1_peer_certificate, s)
SSL_get_peer_certificate(ssl)
```
The use of `__VA_ARGS__` is optional. One could, for example, instead write:
```
#define OSSL_PRAGMA_WRAP(msg) _Pragma(#msg)
#define OSSL_PRAGMA_GCC_WARN(msg) OSSL_PRAGMA_WRAP(GCC warning # msg)
#define OSSL_DEPRECATED_INSTEAD_USE(f1, f2) OSSL_PRAGMA_GCC_WARN(f1 is deprecated use f2 instead) f2
#define SSL_get_peer_certificate(s) OSSL_DEPRECATED_INSTEAD_USE(SSL_get_peer_certificate, SSL_get1_peer_certificate)(s)
SSL_get_peer_certificate(ssl)
```